### PR TITLE
OCPBUGS-32206: [IBU] LVMS volume contents not restored 

### PR DESCRIPTION
**Pull request by leo8a**
Adds support for LVMS in LCA.

The implemented workflow is as follows:
- pre-pivot: LCA patches to `Retain` all PVs created by LVMS and with a Reclaim Policy of `Delete`.
- post-pivot: LCA restores back to `Delete` the Reclaim Policy of all PVs created by LVMS.
- rollback -> idle: LCA restores back to `Delete` the Reclaim Policy of all PVs previously patched.

Also added docs and a couple of entries to the troubleshooting guide.

/cc @browsell @Missxiaoguo @donpenney @jakobmoellerdev

### CODE REVIEW

File: docs/backuprestore-with-oadp.md [hunk 1]
```diff
@@ -256,6 +258,10 @@ status:
     type: Reconciled
 ```
 
+> [!NOTE]
+> When LVMS is used to create PVs dynamically in the cluster, please ensure that the CSI Integration (i.e.,
```
Comment (browsell) [inline docs/backuprestore-with-oadp.md:262]: We can probably do that by default, would not be needed by lso either

File: docs/backuprestore-with-oadp.md [hunk 2]
```diff
@@ -431,7 +437,99 @@ spec:
 > 1. The examples provided are just for reference. Create your own backup and restore CRs based on your needs.
 > 2. The backup and restore CRs must be created in the same namespace where the OADP is installed which is `openshift-adp`.
 > 3. The OADP configmaps should only contain backup and restore CRs which are in YAML format or multiple document YAML format. Other type of CRs are unknown to the LCA and will be ignored.
->
+
+##### 1.1 Backup resource when LVMS is used
+
+When LVMS the selected provider to provide dynamic storage in the cluster, the backup and restore resources for the
```
Comment (browsell) [inline docs/backuprestore-with-oadp.md:443]: nit, reword," When LVMS is used to provide dynamic storage in the cluster "

File: docs/backuprestore-with-oadp.md [hunk 3]
```diff
@@ -431,7 +437,99 @@ spec:
...
+    velero.io/storage-location: default
+  annotations:
+    lca.openshift.io/apply-wave: "2"
+spec:
+  backupName:
+    lvmcluster
+```
+
+> [!IMPORTANT]
+> Mind that `apply-wave` annotation for below object should be numerically minor than the app's one, in this way when
```
[the remark this answers is not in the document]
Comment (leo8a) [reply docs/backuprestore-with-oadp.md]: thanks

File: docs/backuprestore-with-oadp.md (hunk 2 above)
Comment (leo8a) [reply docs/backuprestore-with-oadp.md]: thanks

File: docs/backuprestore-with-oadp.md [hunk 4]
```diff
@@ -431,7 +437,99 @@ spec:
...
+spec:
+  backupName:
+    small-app
+  restorePVs: true               # <- required field
+  restoreStatus:                 # <- required field
+    includedResources:           # <- required field
+      - logicalvolumes           # <- required field
+```
+
+As a complement, we also included below, sample resources to backup and restore the cluster's manifests required
```
Comment (jakobmoellerdev) [inline docs/backuprestore-with-oadp.md:491]: Nit: I think it is important to note that this is always required, not restoring LVMCluster will break LVMS, so not sure if its a "complement"

File: docs/backuprestore-with-oadp.md [hunk 5]
```diff
@@ -431,7 +437,99 @@ spec:
...
+  namespace: openshift-adp
+spec:
+  includedNamespaces:
+  - test
+  includedNamespaceScopedResources:
+  - secrets
+  - persistentvolumeclaims
+  - deployments
+  - statefulsets
+  includedClusterScopedResources:
```
Comment (jakobmoellerdev) [inline docs/backuprestore-with-oadp.md:464]: Nit: it may be worth noting that this is only necessary exactly once per cluster restore since it is cluster scoped. (although specifying this per application is redundant, it also wont break anything)

File: docs/image-based-upgrade.md [hunk 6]
```diff
@@ -149,24 +149,24 @@ LCA will reject the stage transition if it is an invalid transition.
...
-|             |                  |True  |Completed     |False      |Idle, Upgrade    |
-|`Upgrade`    |UpgradeInProgress |True  |InProgress    |False      |Idle             |
-|             |                  |True  |InProgress    |True       |Rollback         |
-|             |UpgradeCompleted  |False |Failed        |False      |Idle             |
-|             |                  |False |Failed        |True       |Rollback         |
-|             |                  |True  |Completed     |True       |Idle, Rollback   |
-|`Rollback`   |RollbackInProgress|True  |InProgress    |True       |N/A              |
-|             |RollbackCompleted |False |Failed        |True       |N/A              |
-|             |                  |True  |Completed     |True       |Idle             |
+| Current Stage | Condition          | Status | Reason         | After Pivot | Valid Next Stages |
```
Comment (jakobmoellerdev) [inline docs/image-based-upgrade.md:152]: Nit: Looks like an accidental format or did something change in the state table?

File: docs/backuprestore-with-oadp.md (hunk 1 above)
Comment (leo8a) [reply docs/backuprestore-with-oadp.md]: gotcha, maybe @jakobmoellerdev can help us confirm this

(keeping the note for now, if / when confirmed I may drop the note - if nothing against it)

**Review by jakobmoellerdev — COMMENTED** [review]
Great resolution of Veleros issues with Backup/Restore of the PVs and double checked it against my OADP Restore POC which looks good. Just minor nits. Thanks for covering this in such detail!

File: docs/backuprestore-with-oadp.md (hunk 1 above)
Comment (jakobmoellerdev) [reply docs/backuprestore-with-oadp.md]: I think restic is only necessary in case we want to copy over the data of the PV into the backup itself. For IBU this should never be necessary since the data is kept on disk and reused after pivot.

File: docs/backuprestore-with-oadp.md (hunk 4 above)
Comment (leo8a) [reply docs/backuprestore-with-oadp.md:491]: totally agree, will move it up in the text and rephrase it to better highlight this, thanks for commenting it

File: docs/image-based-upgrade.md (hunk 6 above)
Comment (leo8a) [reply docs/image-based-upgrade.md:152]: yea, nothing actually changed in the state table... was just my IDE complaining about "bad table formatting"

File: docs/backuprestore-with-oadp.md (hunk 5 above)
Comment (leo8a) [reply docs/backuprestore-with-oadp.md:464]: k, will drop a quick side note for that as well

File: docs/image-based-upgrade.md [hunk 7]
```diff
@@ -324,13 +325,14 @@ Pivot
...
+- Once LCA starts, it will restore the saved IBU CR.
 - Restore the remaining platform configuration.
 - Wait for the platform to recover - Cluster/day2 operators and MCP are stable.
 - Apply extra manifests that were saved pre-pivot.
-- Apply any OADP restore CRs that were saved pre-pivot. Platform artifacts will be restored first including ACM artifacts if the system is managed by ACM.
+- Apply any OADP restore CRs that were saved pre-pivot. Platform artifacts will be restored first, including ACM artifacts if the system is managed by ACM.
+- If LVMS is used, automatically restores the `.spec.persistentVolumeReclaimPolicy` field of PVs to its original (pre-pivot) value.
 
-Upon completion, the condition will be updated to "Upgrade Completed".
+Upon completion, the condition will be updated to "Upgrade Completed."
```
Comment (donpenney) [inline docs/image-based-upgrade.md:335]: nit: I think the original would be correct. Both grammatically and because the condition string doesn't include a period. Mind you, the condition type also wouldn't include a space.

File: docs/troubleshooting.md [hunk 8]
```diff
@@ -47,3 +47,120 @@ dataprotectionapplication-1   Available   33s              8d    true
...
+spec:
+  includedNamespaces:
+  - test
+  includedNamespaceScopedResources:
+  - secrets
+  - persistentvolumeclaims
+  - deployments
+  - statefulsets
+  includedClusterScopedResources:
+  - persistentVolumes              # <- required field
```
Comment (Missxiaoguo) [inline docs/troubleshooting.md:97]: I assume that all PVs will be backed up. If there are PVs created by both LSO and LVMS, will the PVs created by LSO be affected after restore in the post-pivot?

Comment (browsell) [reply docs/troubleshooting.md:97]: We need to update the documentation, only LSO or LVMS can be used, not both

Comment (leo8a) [reply docs/troubleshooting.md:97]: should we need to drop a note for that in this PR, or will it be included later on (in the official docs, for instance)?

Comment (browsell) [reply docs/troubleshooting.md:97]: We should do both

Comment (leo8a) [reply docs/troubleshooting.md:97]: Ok, just dropped [a note for it here](https://github.com/openshift-kni/lifecycle-agent/pull/448/commits/3e5e750549f3ca5cef80217ceb1938f75961a720#diff-d08037f5ce2c67f942ba8bf93d5d9c83711172f9d289737d2a44407044d1a4f8R447-R449)

/cc @amolnar-rh   (keeping you in the loop - for when adding LVMS support to official docs)

File: internal/backuprestore/backup.go [hunk 9]
```diff
@@ -254,6 +254,70 @@ func (h *BRHandler) createNewBackupCr(ctx context.Context, backup *velerov1.Back
...
+		return fmt.Errorf("failed to list PersistentVolumes: %w", err)
+	}
+
+	if len(pvList.Items) == 0 {
+		h.Log.Info("No PersistentVolumes found in the cluster, skipping.")
+		return nil
+	}
+
+	for _, pv := range pvList.Items {
+		hasLvmsAnnotation := pv.GetAnnotations()[topolvmAnnotation] == topolvmValue
```
Comment (Missxiaoguo) [inline internal/backuprestore/backup.go:273]: The annotations could be nil if pv is manually created with `oc create -f  `.
It's safer to add check `if pv.GetAnnotations() != nil`

File: internal/backuprestore/backup.go [hunk 10]
```diff
@@ -254,6 +254,70 @@ func (h *BRHandler) createNewBackupCr(ctx context.Context, backup *velerov1.Back
...
+		return fmt.Errorf("failed to list PersistentVolumes: %w", err)
+	}
+
+	if len(pvList.Items) == 0 {
+		h.Log.Info("No PersistentVolumes found in the cluster, skipping.")
+		return nil
+	}
+
+	for _, pv := range pvList.Items {
+		hasReclaimPolicyAnnotation := pv.GetAnnotations()[updatedReclaimPolicyAnnotation] == "true"
```
Comment (Missxiaoguo) [inline internal/backuprestore/backup.go:304]: same comment as above

**Review by Missxiaoguo — COMMENTED** [review]
Others look good to me!

File: internal/backuprestore/backup.go (hunk 9 above)
Comment (leo8a) [reply internal/backuprestore/backup.go]: nice, thanks
